nvflare/fuel/utils/fobs/fobs.py

```diff
@@ -0,0 +1,182 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import importlib
+import inspect
+import logging
+import os
+from os.path import dirname, join
+from typing import Any, BinaryIO, Dict, Type, Union
+
+import msgpack
+
+from nvflare.fuel.utils.fobs.decomposer import Decomposer
+
+__all__ = [
+    "register",
+    "register_folder",
+    "num_decomposers",
+    "serialize",
+    "serialize_stream",
+    "deserialize",
+    "deserialize_stream",
+]
+
+FOBS_TYPE = "__fobs_type__"
+FOBS_DATA = "__fobs_data__"
+MSGPACK_TYPES = (None, bool, int, float, str, bytes, bytearray, memoryview, list, dict)
+
+log = logging.getLogger(__name__)
+_decomposers: Dict[str, Decomposer] = {}
+_decomposers_registered = False
+
+
+def _get_type_name(cls: Type) -> str:
+    module = cls.__module__
+    if module == "builtins":
+        return cls.__qualname__
+    return module + "." + cls.__qualname__
+
+
+def register(decomposer: Union[Decomposer, Type[Decomposer]]) -> None:
+    """Register a decomposer. It does nothing if decomposer is already registered for the type
+
+    Args:
+        decomposer: The decomposer type or instance
+    """
+
+    name = _get_type_name(decomposer.supported_type())
+    if name in _decomposers:
+        return
+
+    if inspect.isclass(decomposer):
+        instance = decomposer()
+    else:
+        instance = decomposer
+
+    if not isinstance(instance, Decomposer):
+        log.error(f"Class {instance.__class__} is not a decomposer")
+        return
+
+    _decomposers[name] = instance
+
+
+def register_folder(folder: str, package: str) -> None:
+    """Scan the folder and register all decomposers found.
+
+    Args:
+        folder: The folder to scan
+        package: The package to import the decomposers from
+    """
+    for module in os.listdir(folder):
+        if module != "__init__.py" and module[-3:] == ".py":
+            decomposers = package + "." + module[:-3]
+            imported = importlib.import_module(decomposers, __package__)
+            for _, cls_obj in inspect.getmembers(imported, inspect.isclass):
+                if issubclass(cls_obj, Decomposer) and not inspect.isabstract(cls_obj):
+                    register(cls_obj)
+
+
+def _register_decomposers() -> None:
+    global _decomposers_registered
+
+    if _decomposers_registered:
+        return
+
+    register_folder(join(dirname(__file__), "decomposers"), ".decomposers")
+    _decomposers_registered = True
+
+
+def num_decomposers() -> int:
+    """Returns the number of decomposers registered.
+
+    Returns:
+        The number of decomposers
+    """
+    return len(_decomposers)
+
+
+def _fobs_packer(obj: Any) -> Any:
```

Comment:
see from the result you are returning a dict, should the return type be dict?

Reply:
Fixed.